tinybytes/src/lib.rs

```diff
@@ -291,6 +320,128 @@ impl Serialize for Bytes {
     }
 }
 
+pub struct RefCountedCell {
+    raw: RawRefCountedCell,
+}
+
+unsafe impl Send for RefCountedCell {}
+unsafe impl Sync for RefCountedCell {}
+
+impl RefCountedCell {
+    #[inline]
+    /// Creates a new `RefCountedCell` from the given data and vtable.
+    ///
+    /// The data pointer can be used to store arbitrary data, that won't be dropped until the last
+    /// clone to the `RefCountedCell` is dropped.
+    /// The vtable customizes the behavior of a Waker which gets created from a RawWaker. For each
+    /// operation on the Waker, the associated function in the vtable of the underlying RawWaker
+    /// will be called.
+    ///
+    /// # Safety
+    ///
+    /// * The value pointed to by `data` must be 'static + Send + Sync
+    pub const unsafe fn from_raw(data: NonNull<()>, vtable: &'static RefCountedCellVTable) -> Self {
+        RefCountedCell {
+            raw: RawRefCountedCell { data, vtable },
+        }
+    }
+}
+
+impl Clone for RefCountedCell {
+    fn clone(&self) -> Self {
+        unsafe { (self.raw.vtable.clone)(self.raw.data.as_ptr().cast_const()) }
+    }
+}
+
+impl Drop for RefCountedCell {
+    fn drop(&mut self) {
+        unsafe { (self.raw.vtable.drop)(self.raw.data.as_ptr().cast_const()) }
+    }
+}
+
+struct RawRefCountedCell {
+    data: NonNull<()>,
+    vtable: &'static RefCountedCellVTable,
+}
+
+pub struct RefCountedCellVTable {
+    pub clone: unsafe fn(*const ()) -> RefCountedCell,
+    pub drop: unsafe fn(*const ()),
+}
+
+/// Creates a refcounted cell.
+///
+/// The data passed to this cell will only be dopped when the last
+/// clone of the cell is dropped.
+fn make_refcounted<T: Send + Sync + 'static>(data: T) -> RefCountedCell {
+    /// A custom Arc implementation that contains only the strong count
+    ///
+    /// This struct is not exposed to the outside of this functions and is
+    /// only interacted with through the `RefCountedCell` API.
+    struct CustomArc<T> {
+        rc: AtomicUsize,
+        #[allow(unused)]
+        data: T,
+    }
+
+    unsafe fn custom_arc_clone<T>(data: *const ()) -> RefCountedCell {
+        let custom_arc = data as *const CustomArc<T>;
+        let rc = unsafe {
+            std::ptr::addr_of!((*custom_arc).rc)
+                .as_ref()
+                .unwrap_unchecked()
+        };
```

Comment:
When data this is NonNull<()>, we can just idiomatically access it without addr_of stuff:
```
let arc : &CustomArc<T> = data.cast().as_mut();
```